### PR TITLE
Add Committer role to the project

### PR DESCRIPTION
**Pull request by matks**
| Questions     | Answers
| ------------- | -------------------------------------------------------
| Description  | This Pull Request is a proposal to add the role of Committer in the project.<br/><br/>A Committer is a contributor who, after demonstrating his skills and motivation towards helping the review of contributions, would be granted a partial approval power on pull requests for the Core repository. As you can see in the PR content, a committer would be able to approve Pull Requests however a Pull Request would be accepted only if at least one maintainer approved it.<br/><br/>So the current rule is "on the core repository, every pull request needs to be approved by two maintainers".<br/>The rule would become "on the core repository, every pull request needs to be approved by two maintainers or by one committer and one maintainer".<br/><br/>This is really about helping, a reviewer or multiple committers cannot alone get a PR approved and mergeable. Maintainers keep the ability to decide if it’s merged or not as gatekeepers for the project. Becoming a committer could become a first step towards being a maintainer.<br/><br/>I did not include this role on other repositories because other repositories only require one approval so the "partial approval" would not apply.<br/><br/>This is a first step, I think committers could (and should) be granted further rights as we experiment with the concept and trust more and more community people.
| Fixed ticket | 

This suggestion is inspired by [Drupal committers](https://www.drupal.org/community/contributor-guide/find-your-role) and [NodeJS committers](https://medium.com/the-node-js-collection/healthy-open-source-967fa8be7951).

### CODE REVIEW

**Review by NeOMakinG — APPROVED** [review]
I've a question

- Do github permit to have a special role for this, counting the review as a valid review in order to merge the PR?
- Can this role be scoped on different context? (front-end, back-end)

**Comment by matks** [comment]
> * Do github permit to have a special role for this, counting the review as a valid review in order to merge the PR?

Yes, GitHub allows us to configure
- the number of required approvals (= 2 on the main repository)
- at least 1 approval from 1 group (= maintainer)

**Comment by NeOMakinG** [comment]
@Hlavtox @mparvazi 

Can't wait to see you in that review group at least! You deserve it as your review often have a lot of value!

**Comment by matks** [comment]
> @Hlavtox @mparvazi
> 
> Can't wait to see you in that review group at least! You deserve it as your review often has a lot of value!

Indeed we can add them to the list already :) if you agree

**Review by matthieu-rolland — APPROVED** [review]

File: content/maintainers-guide/project-organization.md
Comment (Progi1984) [inline content/maintainers-guide/project-organization.md:45]:
```suggestion
```

Define the role before accepting people in this role.

**Review by Progi1984 — CHANGES_REQUESTED** [review]
The role must be defined (context, missions, acceptance criteria) before accepting new people.

**Comment by matks** [comment]
> The role must be defined (context, missions, acceptance criteria) before accepting new people.

Context and mission: already provided

```
Reviewers help maintainers review code contributions.
```

It is a subset of the context and mission of maintainers:

```
Maintainers provide the technical vision for the project, review code contributions, and ensure that the software is developed according to the product vision.
```

Criteria: this document title is "project organization".

The criteria of maintainers have not been written inside this document. This document also does not contain the criteria to apply as issue managers or any other roles. I do not think the criteria to apply should be written in there.

Maybe criteria should go into a subpage inside Maintainer Guide, as Reviewer role is a subset of Maintainer role?

**Comment by saulaski** [comment]
To keep the documentation simple and lighter to read, perhaps it should only mention "reviewers" alone instead of "maintainers and reviewers" ?
We could indeed consider all maintainers as reviewers by default, as a basic role, but they have extended roles beyond review.

**Comment by NeOMakinG** [comment]
> The role must be defined (context, missions, acceptance criteria) before accepting new people.

I would also say that we should add some roles scopes like Front-end and Back-end

- Back-end stuff in this project means that you master different languages, it's not like you were a JS backend engineer, it means that you are a PHP engineer
- Front-end scope is really big now, it evolved a lot and I really think that It's hard to be "Full stack" in such a project, it means that you are a PHP, JS, and HTML/CSS engineer

**Comment by Hlavtox** [comment]
Also, if such a role is created, it should have some kind of argumentation power.

Now, when I checked one of the maintainers PR and suggested some CSS changes because it was really wrong, he just clicked "Resolved" (=I don't care). 

If Valentin didn't step in and said "Yeah it's really crap", it would get merged.

**Comment by Hlavtox** [comment]
> Front-end scope is really big now, it evolved a lot and I really think that It's hard to be "Full stack" in such a project, it means that you are a PHP, JS, and HTML/CSS engineer

+1 Don't take it wrong, but the quality of HTML/CSS not written by @NeOMakinG is terrible. 😄

File: content/maintainers-guide/project-organization.md
Comment (eternoendless) [inline content/maintainers-guide/project-organization.md:41]:
```suggestion
### Committers

Committers participate in reviewing code contributions and are able to include them in the project once they have been accepted and verified.

Current committers: No one yet.
```

File: content/maintainers-guide/reviewing-pull-requests.md
Comment (eternoendless) [inline content/maintainers-guide/reviewing-pull-requests.md:22]:
```suggestion
Reviewers with merge rights (Committers and Maintainers) are responsible for the Pull Requests they approve. The main objective of this process is to make sure that the change being reviewed does not produce undesired side effects and that its inclusion is coherent with the project's objectives.
```

File: content/maintainers-guide/reviewing-pull-requests.md
Comment (eternoendless) [inline content/maintainers-guide/reviewing-pull-requests.md:56]:
```suggestion
Reviewers might ask questions to the Pull Request author (for example to better understand a technical choice). Simply writing the question as a comment is enough.
```

File: content/maintainers-guide/reviewing-pull-requests.md
Comment (eternoendless) [inline content/maintainers-guide/reviewing-pull-requests.md:58]:
```suggestion
When doing so, it is recommended to add the label "Waiting for author" on the Pull Request. This helps other reviewers to know the state of this Pull Request.
```

File: content/maintainers-guide/reviewing-pull-requests.md
Comment (eternoendless) [inline content/maintainers-guide/reviewing-pull-requests.md:112]:
```suggestion
- The change has been approved by at least one maintainer.
```

File: content/maintainers-guide/reviewing-pull-requests.md
Comment (eternoendless) [inline content/maintainers-guide/reviewing-pull-requests.md:118]:
```suggestion
On the [PrestaShop Core repository][prestashop-core-repository]), Pull Requests must be approved by two reviewers with merge rights, including at least one maintainer. For example:

- One committer and one maintainer
- Two maintainers
```

**Review by eternoendless — COMMENTED** [review]
I suggest changing the "reviewer" title to "committer"

**Review by Progi1984 — CHANGES_REQUESTED** [review]
I disagree. New roles should be added with all informations (acceptance criteri is the most important). You can modify the subpage in this PR.

**Comment by matks** [comment]
Hello everybody,

I tried to meet every feedback

> The role must be defined (context, missions, acceptance criteria) before accepting new people.

I removed the new people. This PR now onlys introduces the Committer role. No candidate is mentioned.

> I would also say that we should add some roles scopes like Front-end and Back-end

So you think people should apply as
- a full-stack committer
or
- a back-end committer
or
- a front-end committer

> Also, if such a role is created, it should have some kind of argumentation power.

We can explore this but I don't think we need to grant such power. If such power is needed, it's because two people of the organization cannot talk together and agree on something. This is very bad. We try to choose candidates that are nice people and are able to talk through such disagreements 😊 . Our Code of Conduct even enforces that we listen to each other. So I don't think we need to add in the role description "what happens if you need to clash". 😅 

> New roles should be added with all informations (acceptance criteria is the most important). You can modify the subpage in this PR.

Upon your suggestion I can add a voting system to grant Committer role to people.

1. A maintainer states he wants to grant Committer role to developer John Doe
2. A vote is opened, it must receive a [simple majority](http://en.euabc.com/word/783) approval to grant the role. Maintainers and committers have one vote.

**Comment by matks** [comment]
🆙 😄

**Comment by kpodemski** [comment]
@matks 

> Upon your suggestion I can add a voting system to grant Committer role to people.
> A maintainer states he wants to grant Committer role to developer John Doe
> A vote is opened, it must receive a [simple majority](http://en.euabc.com/word/783) approval to grant the role. Maintainers and committers have one vote.

Simple majority. Sounds good.

> So you think people should apply as
> 
> a full-stack committer
> or
> a back-end committer
> or
> a front-end committer

I'm not sure about this. Maybe it should state that we can give review permission to one or more repositories based on the reviewer's skillset. Any developer might review some parts of the documentation. Having a split between roles in this manner is overkill IMHO.

**Comment by matks** [comment]
In order not to keep this PR waiting for longer, I can suggest I add the Vote part to this PR and then we move forward and merge this of we got approvals?

So I create a page "How to become committer" similar to https://github.com/PrestaShop/open-source/blob/master/content/maintainers-guide/how-to-become-a-maintainer.md

**Comment by kpodemski** [comment]
@matks 🆗

**Comment by matks** [comment]
Good evening everyone and apologies for the delay. I added a "How to become a committer" page very similar to "How to become a maintainer"

**Comment by matks** [comment]
Ping @kpodemski

File: content/maintainers-guide/how-to-become-a-committer.md
Comment (kpodemski) [inline content/maintainers-guide/how-to-become-a-committer.md:53]:
```suggestion
A project committer can be revoked for any of the following reasons:
```

Comment (matks) [reply content/maintainers-guide/how-to-become-a-committer.md:53]: Thanks for proof reading

**Review by kpodemski — APPROVED** [review]

**Comment by Progi1984** [comment]
@ttoine How and Who can add a new role in the project ?

**Review by matthieu-rolland — APPROVED** [review]

**Review by jolelievre — COMMENTED** [review]
HI @matks 

ok for the content in the pages. However, is it on purpose that a binary `hugo` has been added in this PR?

**Comment by matks** [comment]
> ok for the content in the pages. However, is it on purpose that a binary hugo has been added in this PR?

@jolelievre No 😭 I rebased the PR yesterday and I missed the line because it was so small

**Review by jolelievre — DISMISSED** [review]
Ok so I block it, just in case, until the file is cleared.

**Comment by jolelievre** [comment]
Request changes don't block the merge anyway 😅. Maybe the configuration of this repository needs some love a bit ^^

File: content/maintainers-guide/how-to-become-a-committer.md
Comment (jolelievre) [inline content/maintainers-guide/how-to-become-a-committer.md:8]: Oh actually one comment on this. Should the commiters have merge rights as well? Or only approval rights?

We could keep merge rights for maintainers only as a safeguard. Since anyway at least one approval from a maintainer is required, it makes sense to leave the merge rights to maintainers only? What do you think?

Comment (matks) [reply content/maintainers-guide/how-to-become-a-committer.md:8]: Interesting. In theory anybody could merge a PR as long as it matches our criteria: 2 approvals (at least 1 maintainer), CI passes, verified by QA, and any item like wording or UX validated if eligible so maybe we could also open merge rights.

But I think GitHub will not allow us to do this. GitHub requests a group to be "code owner" and I think this is the people able to push and merge on the branch.

File: content/maintainers-guide/how-to-become-a-committer.md
Comment (Progi1984) [inline content/maintainers-guide/how-to-become-a-committer.md:56]: Please define the lack of activity. It's not clear.

One, Ten, Twenty reviews in three months is a lack of activity ?

File: content/maintainers-guide/reviewing-pull-requests.md
Comment (Progi1984) [inline content/maintainers-guide/reviewing-pull-requests.md:130]:
```suggestion
- The change has received all the required approvals.
```

File: content/maintainers-guide/how-to-become-a-committer.md
Comment (Progi1984) [inline content/maintainers-guide/how-to-become-a-committer.md:8]:
```suggestion
In the PrestaShop project, committer are contributors with approval rights, but not merge rights. 
```

File: content/maintainers-guide/how-to-become-a-committer.md
Comment (Progi1984) [inline content/maintainers-guide/how-to-become-a-committer.md:11]: Please explain differences with maintainer. It's not clear.

File: content/maintainers-guide/how-to-become-a-committer.md
Comment (Progi1984) [inline content/maintainers-guide/how-to-become-a-committer.md:12]: Is that a commiter can become a maintainer ? and how can he do that ?

**Review by Progi1984 — COMMENTED** [review]
Some feedbacks

**Review by Progi1984 — DISMISSED** [review]
Some feedbacks

File: content/maintainers-guide/how-to-become-a-committer.md
Comment (matks) [reply content/maintainers-guide/how-to-become-a-committer.md:56]: I dont think we should put a number here. You cannot measure the activity of a developer by a raw number of reviews. It would be like trying to measure a developer output by the number of lines of code he/she produces.

You can do twenty useless code reviews on 1 day or you can provide 1 deep code review for one PR with high level feedback.

I know you like numbers and rules @Progi1984 but you wont be able to create and define _relevant and consistent_ rules that apply to real world 😄 . The world and human beings are too messy to be able to fit into tight words.

So just like for Code Review is done by a humain brain and not an automated tool, I suggest "lack of activity" is something that should be discussed and evaluated together as maintainer on a case-by-case basis.

If tomorrow for example the maintainer group thinks I, matks, present a lack of activity, it is a topic to be discussed by maintainers. Then if no consensus "yes or no" can be found, then a vote might be done to decide whether or not I present a lack of activity. Only brains can decide whether there is or not a lack of activity, you cant say "lack of activity is less than X reviews per month". Multiple factors have to be acknowledged. Trust the people, not a random metric.

I think we dont clarify every single word. Actually PrestaShop is an one open source project with the _most detailed process to become maintainer_.  We're already way above average. Look at Symfony, Doctrine, Laravel, PHP, jQuery, VueJS, I was not able to find clear and detailed "how to become a maintainer" guidelines for these projects. When you browse the topics about maintainer what you can see is basically "want to become a maintainer? contribute then we'll chat" 😄  . If I look at open source projects of size and maturity comparable to us, we're doing way better.

If you have an example of three other open source projects of same size / maturity that PrestaShop that have defined "lack of activity for a maintainer is less than X code reviews per month" please share 😄

File: content/maintainers-guide/how-to-become-a-committer.md
Comment (MeKeyCool) [reply content/maintainers-guide/how-to-become-a-committer.md:11]: If I understand well, it is like B include A example 



with A = being a committer and B = being a maintainer 
So every responsibility/power that can take a committer can be taken by a maintainer as well but not all responsibilities/powers of maintainers are given to committers.

Does this example make things clearer ?

File: content/maintainers-guide/how-to-become-a-committer.md
Comment (MeKeyCool) [reply content/maintainers-guide/how-to-become-a-committer.md:12]: Do you mean like a specific process for committers to become maintainer that would be easier than standard process to become maintainer ?

**Review by MeKeyCool — COMMENTED** [review]
> > I would also say that we should add some roles scopes like Front-end and Back-end
> 
> So you think people should apply as
> 
>     * a full-stack committer
>       or
> 
>     * a back-end committer
>       or
> 
>     * a front-end committer

Then why not creating ***wording***, ***product*** and ***qa*** scopes as well ? :thinking: 
Probably to early in project, this "scope" idea may be another issue i think cause it would induce lot of debates :thinking:

**Comment by MeKeyCool** [comment]
Will committers allowed to assign tasks from https://github.com/PrestaShop/PrestaShop/projects ?

File: content/maintainers-guide/how-to-become-a-committer.md
Comment (Progi1984) [reply content/maintainers-guide/how-to-become-a-committer.md:56]: If we can't explain what is lack of activity, we should remove it.

```suggestion
```

Comment (matks) [reply content/maintainers-guide/how-to-become-a-committer.md:56]: Hello ! This weekend I found another example about a "broad" maintainer process: https://make.wordpress.org/core/2020/02/09/what-does-it-mean-to-be-a-component-maintainer-a-refresher/

This is a project that we can relate to in terms of size and maturity (although I think they are actually even better than us) and as you can see they have not written rules for everything. The "how to become a maintainer" list is simply "Contact us". No statement about requirements, no statement about "what can make you lose the status"

When I read this I really think we are over-engineering this process. I think we can stay with this "Lack of activity" sentence and say that when we think there is indeed an issue we discuss it, the full maintainer group.

But that is my opinion and I respect that you have another opinion.

This PR has been open for 2 months now and I believe most maintainers agree to it so I merge the PR but we can continue discussing, with the group, this topic and see what the group thinks about this line.
-> https://github.com/PrestaShop/open-source/issues/95

**Comment by matks** [comment]
Hi everyone,

As 6 people (a majority of maintainers) have approved the PR (me, Jonathan, Matthieu, Valentin, Pablo and Krystian) I move forward and merge the PR, but [I also open](https://github.com/PrestaShop/open-source/issues/95) a GitHub issue to handle the questions of about the activity line.

File: content/maintainers-guide/how-to-become-a-committer.md
Comment (matks) [reply content/maintainers-guide/how-to-become-a-committer.md:11]: @Progi1984 Difference is in the approval rights :) this line is about "mission". Both committers and maintainers have the same mission about code review, not the same rights scope.

Just like I opened https://github.com/PrestaShop/open-source/issues/95 if you feel like there is a need for improvement, let's discuss it with the whole group 👍

File: content/maintainers-guide/how-to-become-a-committer.md
Comment (matks) [reply content/maintainers-guide/how-to-become-a-committer.md:12]: @Progi1984 Yes. If this document does not say "committers cannot become maintainers", it means that it's possible.

File: content/maintainers-guide/how-to-become-a-committer.md
Comment (matks) [reply content/maintainers-guide/how-to-become-a-committer.md:56]: I resolve the conversation in the scope of this issue but I'll be happy to discuss this matter with the whole group in https://github.com/PrestaShop/open-source/issues/95

**Pull request merged**
